### DISCUSSION TITLE
Lesson 6 : List index out of range (MacOS Catalina 10.15.7)

### DISCUSSION THREAD

**Discussion by JamiDEV** [smartcontractkit/full-blockchain-solidity-course-py#451] [Q&A] [answered]
Lesson 6 has been really beating me down. 

another major issue,  for the life of me I can't understand why, I've checked the code and it all seems correct what am I missing?  Mind you when I deploy without targeting a network it runs just fine. 

 the error:
![Screen Shot 2021-11-27 at 10 51 53 PM](https://user-images.githubusercontent.com/53233466/143731495-2c26eb63-5724-48b5-9504-3432801af7b7.png)

i'm looking for the for brownie/network/contract.py file but can't find it, would this be somewhere else?  I do understand the error, I've tried to change the fund me parameter from -1 to 0, but nothing works.

**Comment by cromewar** [comment]
Hello @JamiDEV when brownie is looking for `FundMe[-1]` it means is trying to get the json info of that contract after the last deployment, so the problem might be to due deployment, did you deploy `FundMe.sol` contract after trying to access to it?

**Reply by JamiDEV** [reply]
not sure what you mean by deploy FundMe.sol. The FundMe contract is below. 


```Solidity
pragma solidity ^0.6.6;

import "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol";
import "@chainlink/contracts/src/v0.6/vendor/SafeMathChainlink.sol";

contract FundMe {

mapping(address => uint256) public addressToAmountFunded;
address[] public funders;
address public owner;
AggregatorV3Interface public priceFeed;



 constructor(address _priceFeed) public {
     priceFeed = AggregatorV3Interface(_priceFeed);
     owner = msg.sender;

 }

    function fund() public payable {
        uint256 mimimumUSD = 50 * 10**18;
        require(
            getConversionRate(msg.value) >= mimimumUSD,
            "You need to spend more ETH!"
        );
        addressToAmountFunded[msg.sender] += msg.value;
        funders.push(msg.sender);
    }
    function getVersion() public view returns (uint256) {
        return priceFeed.version();
    }
    function getPrice() public view returns (uint256) {
        (, int256 answer, , , ) = priceFeed.latestRoundData();
        return uint256(answer * 10000000000);
    }
    // 1000000000
    function getConversionRate(uint256 ethAmount)
        public
        view
        returns (uint256)
    {
        uint256 ethPrice = getPrice();
        uint256 ethAmountInUsd = (ethPrice * ethAmount) / 1000000000000000000;
        return ethAmountInUsd;
    }
    function getEntranceFee() public view returns (uint256) {
        // mimimumUSD
        uint256 mimimumUSD = 50 * 10**18;
        uint256 price = getPrice();
        uint256 precision = 1 * 10**18;
        return (mimimumUSD * precision) / price;
    }
    modifier onlyOwner() {
        require(msg.sender == owner);
        _;
    }
    function withdraw() public payable onlyOwner {
        msg.sender.transfer(address(this).balance);
        for (
            uint256 funderIndex = 0;
            funderIndex < funders.length;
            funderIndex++
        ) {
            address funder = funders[funderIndex];
            addressToAmountFunded[funder] = 0;
        }
        funders = new address[](0);
    }
}
```

**Reply by cromewar** [reply]
I mean you would have to compile `brownie compile` in order to your deploy script could access to `FundMe` file. 

```Python
from brownie import FundMe
from scripts.helpful_scripts import get_account


def fund():
    fund_me = FundMe[-1]
```

This code takes the last `FundMe` deployment, so you need to compile it first.

**Reply by JamiDEV** [reply]
I did that and still didn't work, I'm wondering if it's an issue with the FundMe. sol file? 


![Screen Shot 2021-11-28 at 10 09 45 PM](https://user-images.githubusercontent.com/53233466/143812310-a9b1fed0-d282-46a0-861e-4e5c2d1bd831.png)

**Reply by cromewar** [reply]
I reviewed your code and everything is ok, could you please share for deploy script code, and also a screenshot of your project folder structure? additionally try to run the script without the --network prefix

**Reply by cromewar** [reply]
Also you can add the following to `fund_and_withdraw` code:

```python
from scripts.deploy import deploy_fund_me
 
def main():
    deploy_fund_me()
    fund()
    withdraw()

```

**Reply by JamiDEV** [reply]
my folder scruture 
![Screen Shot 2021-11-30 at 11 54 51 PM](https://user-images.githubusercontent.com/53233466/144187124-fe7e6e8d-fd77-4bd3-94f4-0ef8e1e16f95.png)

so I added the deploy_fund_me() function to my main and this happened 


![Screen Shot 2021-12-01 at 12 02 09 AM](https://user-images.githubusercontent.com/53233466/144187281-cdf7fec6-1224-46e7-938f-2f558fcae014.png)

**Reply by cromewar** [reply]
Try this.

1. on vscode terminal type: `rm -rf build/` 
2. try again with `brownie run.... `

**Reply by JamiDEV** [reply]
it did the same thing but the last line reads

Gas estimation failed: 'execution reverted: You need to spend more ETH!'. This transaction will likely revert. If you wish to broadcast, you must set the gas limit manually.

**Reply by cromewar** [reply]
According to official docs: https://eth-brownie.readthedocs.io/en/stable/api-network.html you could increase the gas_limit() in order to ensure your transaction will not revert.

**Reply by JamiDEV** [reply]
gas_limit() where? In what file would that be in?

**Reply by cromewar** [reply]
is a parameter you can add on brownie after `{"from": account, }`also I noticed the error is saying you need to spend more eth, so make sure your  functions are something like this:

```Python

def fund():
    fund_me = FundMe[-1]
    account = get_account()
    entrance_fee = fund_me.getEntranceFee()
    print(entrance_fee)
    print(f"The current entry fee is {entrance_fee}")
    print("Funding")
    fund_me.fund({"from": account, "value": entrance_fee})

```

on the contract
 ```Solidity
   function fund() public payable {
        uint256 mimimumUSD = 50 * 10**18;
        require(
            getConversionRate(msg.value) >= mimimumUSD,
            "You need to spend more ETH!"
        );
        addressToAmountFunded[msg.sender] += msg.value;
        funders.push(msg.sender);
    }
```

also you can add a little more on the entrance fee to be sure you are sending the correct amount like: 

fund_me.fund({"from": account, "value": entrance_fee+ addiotionalAmountYouwant})

**Reply by JamiDEV** [reply]
Okay.  I think I'm just gonna move on, I understand the issue but now I'm spending more time troubleshooting than learning. I've learned a great deal though.

**Reply by cromewar** [reply]
Feel free to make any other question.

**Reply by JamiDEV** [reply]
thank you brother, you are an immense help.

**Reply by JamiDEV** [reply]
Heads up, your suggestion for adding more money to the entrance fee worked. Patrick had it later in the tutorial.

**Comment by JamiDEV** [comment, marked as answer]
from @cromewar 
> is a parameter you can add on brownie after `{"from": account, }`also I noticed the error is saying you need to spend more eth, so make sure your functions are something like this:
> 
> ```python
> def fund():
>     fund_me = FundMe[-1]
>     account = get_account()
>     entrance_fee = fund_me.getEntranceFee()
>     print(entrance_fee)
>     print(f"The current entry fee is {entrance_fee}")
>     print("Funding")
>     fund_me.fund({"from": account, "value": entrance_fee})
> ```
> 
> on the contract
> 
> ```solidity
>   function fund() public payable {
>        uint256 mimimumUSD = 50 * 10**18;
>        require(
>            getConversionRate(msg.value) >= mimimumUSD,
>            "You need to spend more ETH!"
>        );
>        addressToAmountFunded[msg.sender] += msg.value;
>        funders.push(msg.sender);
>    }
> ```
> 
> also you can add a little more on the entrance fee to be sure you are sending the correct amount like:
> 
> fund_me.fund({"from": account, "value": entrance_fee+ addiotionalAmountYouwant})
